s2/region_term_indexer.go

```diff
@@ -0,0 +1,441 @@
+	rti.regionCoverer.MaxCells = rti.options.maxCells
+
+	covering := rti.regionCoverer.Covering(region)
+	return rti.GetIndexTermsForCanonicalCovering(covering, prefix)
+}
+
+func (rti *RegionTermIndexer) GetIndexTermsForCanonicalCovering(
+	covering CellUnion, prefix string) []string {
+	// See the top of this file for an overview of the indexing strategy.
+	//
+	// Cells in the covering are normally indexed as covering terms.  If we are
+	// optimizing for query time rather than index space, they are also indexed
+	// as ancestor terms (since this lets us reduce the number of terms in the
+	// query).  Finally, as an optimization we always index true_max_level()
+	// cells as ancestor cells only, since these cells have the special property
+	// that query regions will never contain a descendant of these cells.
+	var rv []string
+	prevID := CellID(0)
+	tml := rti.options.trueMaxLevel()
+
+	for _, cellID := range covering {
+		level := cellID.Level()
+		if level < tml {
+			rv = append(rv, rti.GetTerm(COVERING, cellID, prefix))
+		}
+
+		if level == tml || !rti.options.optimizeSpace {
+			rv = append(rv, rti.GetTerm(ANCESTOR, cellID.Parent(level), prefix))
+		}
+
+		for (level - rti.options.levelMod) >= rti.options.minLevel {
+			level -= rti.options.levelMod
+			ancestorID := cellID.Parent(level)
+			if prevID != CellID(0) && prevID.Level() > level &&
+				prevID.Parent(level) == ancestorID {
+				break
+			}
+			rv = append(rv, rti.GetTerm(ANCESTOR, ancestorID, prefix))
+		}
+		prevID = cellID
+	}
+
+	return rv
+}
+
+func (rti *RegionTermIndexer) GetQueryTermsForPoint(p Point, prefix string) []string {
+	cellID := cellIDFromPoint(p)
+	var rv []string
+
+	level := rti.options.trueMaxLevel()
+	rv = append(rv, rti.GetTerm(ANCESTOR, cellID.Parent(level), prefix))
+	if rti.options.pointsOnly {
+		return rv
+	}
+
+	for level >= rti.options.minLevel {
+		rv = append(rv, rti.GetTerm(COVERING, cellID.Parent(level), prefix))
+		level -= rti.options.levelMod
+	}
+
+	return rv
+}
+
+func (rti *RegionTermIndexer) GetQueryTermsForRegion(region Region,
+	prefix string) []string {
+	rti.regionCoverer.LevelMod = rti.options.levelMod
+	rti.regionCoverer.MaxLevel = rti.options.maxLevel
+	rti.regionCoverer.MinLevel = rti.options.minLevel
+	rti.regionCoverer.MaxCells = rti.options.maxCells
+
+	covering := rti.regionCoverer.Covering(region)
+	return rti.GetQueryTermsForCanonicalCovering(covering, prefix)
+
+}
+
+func (rti *RegionTermIndexer) GetQueryTermsForCanonicalCovering(
+	covering CellUnion, prefix string) []string {
+	var rv []string
+	prevID := CellID(0)
+	tml := rti.options.trueMaxLevel()
+	for _, cellID := range covering {
+		level := cellID.Level()
+		rv = append(rv, rti.GetTerm(ANCESTOR, cellID, prefix))
+
+		if rti.options.pointsOnly {
+			continue
+		}
+
+		if rti.options.optimizeSpace && level < tml {
+			rv = append(rv, rti.GetTerm(COVERING, cellID, prefix))
+		}
+
+		for level-rti.options.levelMod >= rti.options.minLevel {
```

Comment:
for ; level-rti.options.levelMod >= rti.options.minLevel ; level -= rti.options.levelMod {